src/main/java/plugily/projects/villagedefense/user/data/FileStats.java

```diff
@@ -18,47 +18,102 @@
 
 package plugily.projects.villagedefense.user.data;
 
+import org.bukkit.Bukkit;
 import org.bukkit.configuration.file.FileConfiguration;
+import org.bukkit.entity.Player;
+import org.bukkit.scheduler.BukkitTask;
 import pl.plajerlair.commonsbox.minecraft.configuration.ConfigUtils;
+import pl.plajerlair.commonsbox.sorter.SortUtils;
 import plugily.projects.villagedefense.Main;
 import plugily.projects.villagedefense.api.StatsStorage;
 import plugily.projects.villagedefense.user.User;
+import plugily.projects.villagedefense.utils.LegacyDataFixer;
 import plugily.projects.villagedefense.utils.constants.Constants;
 
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.UUID;
+import java.util.concurrent.atomic.AtomicBoolean;
+
 /**
  * Created by Tom on 17/06/2015.
  */
-public class FileStats implements UserDatabase {
+public class FileStats implements UserDatabase, Runnable {
 
   private final Main plugin;
   private final FileConfiguration config;
+  private final BukkitTask updateTask;
+  private final AtomicBoolean updateRequired = new AtomicBoolean(false);
 
   public FileStats(Main plugin) {
     this.plugin = plugin;
-    config = ConfigUtils.getConfig(plugin, Constants.Files.STATS.getName());
+    new LegacyDataFixer(plugin);
+    this.config = ConfigUtils.getConfig(plugin, Constants.Files.STATS.getName());
+    this.updateTask = Bukkit.getScheduler().runTaskTimerAsynchronously(plugin, this, 40, 40);
   }
 
   @Override
   public void saveStatistic(User user, StatsStorage.StatisticType stat) {
     config.set(user.getUniqueId().toString() + "." + stat.getName(), user.getStat(stat));
-    ConfigUtils.saveConfig(plugin, config, Constants.Files.STATS.getName());
+    updateRequired.set(true);
   }
 
   @Override
   public void saveAllStatistic(User user) {
+    updateStats(user);
+    updateRequired.set(true);
+  }
+
+  @Override
+  public void loadStatistics(User user) {
+    for(StatsStorage.StatisticType stat : StatsStorage.StatisticType.values()) {
+      user.setStat(stat, config.getInt(user.getUniqueId().toString() + "." + stat.getName(), 0));
+    }
+  }
+
+  @Override
+  public Map<UUID, Integer> getStats(StatsStorage.StatisticType stat) {
+    Map<UUID, Integer> stats = new TreeMap<>();
+    for(String string : config.getKeys(false)) {
+      if(string.equals("data-version")) {
+        continue;
+      }
+      stats.put(UUID.fromString(string), config.getInt(string + "." + stat.getName()));
```

Comment:
It may not necessarily happen, but at some point it may be that the UUID conversion from string may have problems, such as the string length may not contain 36 characters or may contain invalid characters. To eliminate this, you need the `IllegalArgumentException` exception handling.